pjsip/src/pjsua-lib/pjsua_pres.c

```diff
@@ -2314,6 +2315,21 @@ static pj_bool_t unsolicited_mwi_on_rx_request(pjsip_rx_data *rdata)
         return PJ_FALSE;
     }
 
+    /* Find which account for the incoming request. */
+    acc_id = pjsua_acc_find_for_incoming(rdata);
+    if (acc_id == PJSUA_INVALID_ID) {
+        PJ_LOG(2, (THIS_FILE,
+                   "Unable to process incoming message %s "
+                   "due to no available account",
+                   pjsip_rx_data_get_info(rdata)));
+
+        pjsip_endpt_respond_stateless(pjsua_var.endpt, rdata,
```

Comment:
temporarily unavailable seems too vague and suggests a temporary error.
Perhaps 481 or 404?

Reply:
> temporarily unavailable seems too vague and suggests a temporary error. Perhaps 481 or 404?

The same error (PJSIP_SC_TEMPORARILY_UNAVAILABLE) was returned on https://github.com/pjsip/pjproject/blob/master/pjsip/src/pjsua-lib/pjsua_pres.c#L839. We should change that as well

Reply:
The 481 response is `481 Subscription does not exist`, so it cannot be used for https://github.com/pjsip/pjproject/blob/master/pjsip/src/pjsua-lib/pjsua_pres.c#L839 .

While here's the recommendation of the RFC https://www.rfc-editor.org/rfc/rfc3265#section-3.2:
```
If any non-SUBSCRIBE mechanisms are defined to create subscriptions,
   it is the responsibility of the parties defining those mechanisms to
   ensure that correlation of a NOTIFY message to the corresponding
   subscription is possible.  Designers of such mechanisms are also
   warned to make a distinction between sending a NOTIFY message to a
   subscriber who is aware of the subscription, and sending a NOTIFY
   message to an unsuspecting node.  The latter behavior is invalid, and
   MUST receive a "481 Subscription does not exist" response (unless
   some other 400- or 500-class error code is more applicable), as
   described in [section 3.2.4](https://www.rfc-editor.org/rfc/rfc3265#section-3.2.4).  In other words, knowledge of a
   subscription must exist in both the subscriber and the notifier to be
   valid, even if installed via a non-SUBSCRIBE mechanism.
```

Reply:
We need the status text "Subscription does not exist", otherwise the default will be "Call/Transaction Does Not Exist" which is not quite appropriate here.

Reply:
As this is a part of the feature called "unsolicited MWI", IMO it is rather ambiguous to respond with "Subscription does not exist" because that's exactly the purpose at the first place. Moreover 4xx is more for UAC-failure, I guess we better use UAS-failure/5xx response code as the error is on UAS, for example `PJSIP_SC_INTERNAL_SERVER_ERROR` or `PJSIP_SC_SERVICE_UNAVAILABLE`?

Reply:
Fair point, I considered that too. That's why I included 404 as alternative and quoted the entire paragraph of the RFC so we can get a more complete understanding.

According to my interpretation, the unsolicited MWI falls under "the non-SUBSCRIBE mechanisms defined to create subscriptions". The fact that there's no SUBSCRIBE message exchanged doesn't mean that there's no subscription, as highlighted again in the closing of the paragraph `subscription must exist in both the subscriber and the notifier to be valid, even if installed via a non-SUBSCRIBE mechanism.`

So here IMO, we rely on the existence of the account for a subscription to exist and be valid.

And for the error code, even though it recommends 481, it doesn't strictly prohibit any other code, so I'm okay with that. But I don't agree that this is a UAS failure since it's also likely that the UAC sends the NOTIFY to the wrong address/account.

Reply:
> According to my interpretation, the unsolicited MWI falls under "the non-SUBSCRIBE mechanisms defined to create subscriptions". The fact that there's no SUBSCRIBE message exchanged doesn't mean that there's no subscription, as highlighted again in the closing of the paragraph `subscription must exist in both the subscriber and the notifier to be valid, even if installed via a non-SUBSCRIBE mechanism.`

This "unsolicited" is really no subscription (not just no SUBSCRIBE, yes I am aware of this too, call transfer is an example). Please check #982 where this features is implemented, it mentions that "unsolicited means NOTIFY is sent without subscription". Also as shown in the code, `mwi_info.evsub` for the callback is set to `NULL` because it does not use subscription (yes, violating the standard).

Reply:
Well, I guess we disagree on this one. To me, there is a subscriber, there is a notifier, and there is a subscription, albeit implicit.

But since the theoretical concept seems to be irrelevant here, I think we'll just move on.

The point is what error code we should send to the remote. I believe the error message should reflect exactly what happened, which is that the "Account does not exist/not found", and not that there's some random internal error. So any error response that conveys that message should be fine. Some alternatives:
- 481 with status "Account does not exist"
- 404 "Account not found"
- or perhaps 604 "Does Not Exist Anywhere"
``` 
   The server has authoritative information that the user indicated in
   the Request-URI does not exist anywhere.
```

Reply:
Changing the message to represent the actual problem (no account to handle) sounds good to me.